Set vf mac

ajitkhaparde opened the pull request.

stephen-hurd commented on drivers/net/bnxt/bnxt_hwrm.c:
```diff
@@ -2415,6 +2415,36 @@ int bnxt_vf_default_vnic_count(struct bnxt *bp, uint16_t vf)
+		return -ENOMEM;
+	}
+	rc = bnxt_hwrm_send_message(bp, &req, sizeof(req));
+	HWRM_CHECK_RESULT;
```
I don't think you can use HWRM_CHECK_RESULT if positive non-zero return values indicate success.

ajitkhaparde replied:
In this case, will bnxt_hwrm_send_message() return a positive non-zero value?
It is used like in any other place as earlier code.

stephen-hurd replied:
bnxt_hwrm_send_message() returns -1 on failure, zero on success.

HWRM_CHECK_RESULT will cause the function it's in to return a positive error code from the response on failure (the error_code value from the response).

See bnxt_hwrm_func_qcfg_current_vf_vlan() for a function that returns a positive value on success.

stephen-hurd commented on drivers/net/bnxt/bnxt_hwrm.c:
```diff
@@ -2415,6 +2415,36 @@ int bnxt_vf_default_vnic_count(struct bnxt *bp, uint16_t vf)
+	if (req.vnic_id_tbl_addr == 0) {
+		RTE_LOG(ERR, PMD,
+		"unable to map VNIC ID table address to physical memory\n");
+		rte_free(vnic_ids);
```
This likely shouldn't be free()d in here.

ajitkhaparde replied:
Let me do a single point of rte_free() in the calling function.

stephen-hurd commented on drivers/net/bnxt/bnxt_hwrm.c:
```diff
@@ -2415,6 +2415,36 @@ int bnxt_vf_default_vnic_count(struct bnxt *bp, uint16_t vf)
+	HWRM_PREP(req, FUNC_VF_VNIC_IDS_QUERY, -1, resp_vf_vnic_ids);
+
+	req.vf_id = rte_cpu_to_le_16(bp->pf.first_vf_id + vf);
+	req.max_vnic_id_cnt = rte_cpu_to_le_32(bp->pf.total_vnics);
```
So we're assuming the size of the passed-in buffer?  Maybe we should have a separate parameter to define the size.

stephen-hurd commented on drivers/net/bnxt/rte_pmd_bnxt.c:
```diff
@@ -450,5 +451,81 @@ int rte_pmd_bnxt_get_vf_tx_drop_count(uint8_t port, uint16_t vf_id, uint64_t *co
+			return 0;
+		}
+	}
+	filter = bnxt_alloc_filter(bp);
```
Where do these get free()d?

ajitkhaparde replied:
There is no delete happening in the VFd, like an add.
I will add something in the course of normal cleanup?

stephen-hurd replied:
VFd attempts to the filters when the VF is deleted by calling set_vf_rx_mac() with the on parameter set to false.  The actual implementation in sriov.c is clearly broken however and should likely be fixed.

But yes, adding cleanup to bnxt_free_all_filters() should be enough.

stephen-hurd commented on drivers/net/bnxt/rte_pmd_bnxt.c:
```diff
@@ -450,5 +451,81 @@ int rte_pmd_bnxt_get_vf_tx_drop_count(uint8_t port, uint16_t vf_id, uint64_t *co
+		goto free;
+
+	memset(&vnic, 0, sizeof(struct bnxt_vnic_info));
+	vnic.fw_vnic_id = rte_le_to_cpu_16(vnic_ids[0]);
```
You should use the default one, not the first one.

stephen-hurd reviewed (commented):
I'm most worried about the cleanup question.

ajitkhaparde committed:
bnxt: add function to add MAC filter for a VF

Everytime a MAC is configured for a VF by the PF, query the VNIC
ID for the VF, and add an L2 filter for the specified MAC address.

Signed-off-by: Ajit Khaparde <ajit.khaparde@broadcom.com>

stephen-hurd commented on drivers/net/bnxt/bnxt_hwrm.c:
```diff
@@ -2500,3 +2500,57 @@ int bnxt_hwrm_func_cfg_vf_set_vlan_anti_spoof(struct bnxt *bp, uint16_t vf)
+	return rte_le_to_cpu_32(resp->vnic_id_cnt);
+}
+
+int bnxt_hwrm_func_qcfg_vf_dflt_vnic_id(struct bnxt *bp, int vf)
```
This is the one that doesn't currently work, right?

Having this call bnxt_hwrm_func_vf_vnic_query() then loop through the vnics should work, then just this one place would need to be updated in the future.

ajitkhaparde replied:
Done.

stephen-hurd commented on drivers/net/bnxt/rte_pmd_bnxt.c:
```diff
@@ -452,3 +453,78 @@ int rte_pmd_bnxt_get_vf_tx_drop_count(uint8_t port, uint16_t vf_id, uint64_t *co
+	rc = bnxt_hwrm_func_vf_vnic_query(bp, vf_id, vnic_ids);
+	if (rc <= 0)
+		goto exit;
+	num_vnic_ids = rc;
```
All of this could go in bnxt_hwrm_func_qcfg_vf_dflt_vnic_id() instead.

stephen-hurd reviewed (approved):
I suppose it's good enough for now.

ajitkhaparde committed:
bnxt: set the mac filter for VF only if default vnic is available

Set the mac filter for VF only if default vnic is available.
Otherwise return an error.

Signed-off-by: Ajit Khaparde <ajit.khaparde@broadcom.com>

ajitkhaparde committed:
bnxt: use bnxt_hwrm_func_qcfg_vf_dflt_vnic_id to fet default VNIC id

Move the code workaround which iterates to find the default VNIC id
to bnxt_hwrm_func_qcfg_vf_dflt_vnic_id()

Signed-off-by: Ajit Khaparde <ajit.khaparde@broadcom.com>
